File: compiler/src/main/java/com/aaroncoplan/waterfall/compiler/statements/helpers/TranslatableStatement.java
package com.aaroncoplan.waterfall.compiler.statements.helpers;

import org.antlr.v4.runtime.ParserRuleContext;

public abstract class TranslatableStatement {

    private final SourcePosition sourcePosition;


    public TranslatableStatement(String filePath, ParserRuleContext parserRuleContext) {
        this.sourcePosition = new SourcePosition(filePath, parserRuleContext.start.getLine(), parserRuleContext.start.getCharPositionInLine());
    }

    public abstract VerificationResult verify();
    public abstract String translate();

    public SourcePosition getSourcePosition() {
        return sourcePosition;
    }
}

File: compiler/src/main/kotlin/com/aaroncoplan/waterfall/compiler/statements/helpers/TranslatableStatement.kt
package com.aaroncoplan.waterfall.compiler.statements.helpers

import org.antlr.v4.runtime.ParserRuleContext

abstract class TranslatableStatement(
    filePath: String,
    parserRuleContext: ParserRuleContext
) : Translatable {

    private val sourcePosition: SourcePosition = SourcePosition(
        filePath,
        parserRuleContext.start.line,
        parserRuleContext.start.charPositionInLine
    )

    fun getSourcePosition(): SourcePosition = sourcePosition
}
